packages/smart-wallet/src/smartWallet.js

```diff
@@ -418,29 +437,26 @@ export const prepareSmartWallet = (baggage, shared) => {
         /** @type {(purse: ERef<RemotePurse>) => Promise<void>} */
         async watchPurse(purseRef) {
           const { address } = this.state;
+          const { helper } = this.facets;
 
           const purse = await purseRef; // promises don't fit in durable storage
 
-          const { helper } = this.facets;
-          // publish purse's balance and changes
-          void E.when(
-            E(purse).getCurrentAmount(),
-            balance => helper.updateBalance(purse, balance),
-            err =>
-              console.error(
-                address,
-                'initial purse balance publish failed',
-                err,
-              ),
-          );
-          void observeNotifier(E(purse).getCurrentAmountNotifier(), {
-            updateState(balance) {
-              helper.updateBalance(purse, balance);
-            },
-            fail(reason) {
-              console.error(address, `failed updateState observer`, reason);
-            },
-          });
+          // This would seem to fit the observeNotifier() pattern,
+          // but purse notifiers are not (always) durable.
+          // If there is an error due to upgrade, retry watchPurse().
+          const notifier = E(purse).getCurrentAmountNotifier();
+          try {
+            // eslint-disable-next-line no-await-in-loop
+            for await (const newBalance of subscribeLatestSimple(notifier)) {
+              helper.updateBalance(purse, newBalance);
+            }
+          } catch (err) {
+            if (isUpgradeDisconnection(err)) {
+              helper.watchPurse(purse); // retry
```

Comment:
No `return`? Not sure who is the initial caller but this looks like a tail recursion, and it feels weird without a return.

Reply:
Actually thinking about this more, this seem too simplistic, and doesn't handle the same disconnection error being replayed, which is why tools like `reconnectAsNeeded` exist which check the incarnation in the disconnect error.

Reply:
I think the problem is in case of a proxy between the notifier in the upgraded vat, and the subscriber. It's possible that this is not the case here, but you could get in an infinite loop if the subscription request somehow rejects with the same disconnect error. The increasing incarnation number is there to help differentiate the case of replayed disconnect error, and should be checked when re-attempting subscription.

Reply:
> No return? Not sure who is the initial caller but this looks like a tail recursion, and it feels weird without a return.

Agreed. https://github.com/Agoric/agoric-sdk/pull/8573/ has a return.

@dckc which PR should be reviewed? I support this one first and the port be in draft until this lands, so that it can match whatever changes are made here